src/frontend/src/binder/relation/mod.rs

```diff
@@ -239,7 +242,19 @@ impl Binder {
         if !has_schema_name
             && let Some(bound_query) = self.cte_to_relation.get(&table_name)
         {
-            let (query, alias) = bound_query.clone();
+            let (query, mut original_alias) = bound_query.clone();
+            debug_assert_eq!(original_alias.name.value, table_name); // The original CTE alias ought to be its table name.
+
+            if let Some(from_alias) = alias {
+                original_alias.name = from_alias.name.clone();
+                let mut alias_iter = from_alias.columns.iter();
+                original_alias.columns = original_alias.columns.iter().map(|ident| {
+                    alias_iter
+                        .next()
+                        .unwrap_or(ident).clone()
+                }).collect();
```

Comment:
nit: `clone` can be avoided here
```suggestion
                original_alias.name = from_alias.name;
                let mut alias_iter = from_alias.columns.into_iter();
                original_alias.columns = original_alias.columns.into_iter().map(|ident| {
                    alias_iter
                        .next()
                        .unwrap_or(ident)
                }).collect();
```